### PR TITLE
BUG: Prevent stackoverflow on self-containing arrays

### PR DESCRIPTION
**Pull request by eric-wieser**
Approaches #8306. This still errors, but it does so at a python level, without
a segfault.

This makes the following not segfault:

```
a = np.array([None])
a[0] = a
bool(a)
```

On the other hand, it still behaves incorrectly

```
In [8]: bool(a)
---------------------------------------------------------------------------
RecursionError                            Traceback (most recent call last)
RecursionError: maximum recursion depth exceeded while converting array to bool

During handling of the above exception, another exception occurred:

SystemError                               Traceback (most recent call last)
<ipython-input-8-7af9bc0b6a67> in <module>()
----> 1 bool(a)

SystemError: <class 'bool'> returned a result with an error set
```

But that's a different bug (#9078)

### CODE REVIEW

File: numpy/core/src/multiarray/number.c [hunk 1]
```diff
@@ -786,10 +786,15 @@ static int
 _array_nonzero(PyArrayObject *mp)
 {
     npy_intp n;
+    int res;
```
Comment (juliantaylor) [inline numpy/core/src/multiarray/number.c]: can be put into the lower if scope

File: numpy/core/src/multiarray/number.c [hunk 2]
```diff
@@ -786,10 +786,15 @@ static int
...
 
     n = PyArray_SIZE(mp);
     if (n == 1) {
-        return PyArray_DESCR(mp)->f->nonzero(PyArray_DATA(mp), mp);
+        if (Py_EnterRecursiveCall(" while converting array to bool"))
```
Comment (juliantaylor) [inline numpy/core/src/multiarray/number.c]: needs curly braces

Comment (eric-wieser) [inline numpy/core/src/multiarray/number.c]: Oops, was just copying cpython there

**Comment by eric-wieser** [comment]
@juliantaylor: Fixed

**Comment by juliantaylor** [comment]
needs a test, then lgtm

**Comment by eric-wieser** [comment]
Tests added - seems that we weren't testing `bool(arr)` at all

File: numpy/core/tests/test_multiarray.py [hunk 3]
```diff
@@ -6404,6 +6404,28 @@ def test_array_scalar_relational_operation(self):
...
+
+        assert_raises(NotImplementedError, bool, np.array(NotConvertible()))
+        assert_raises(NotImplementedError, bool, np.array([NotConvertible()]))
+
+        self_containing = np.array([None])
+        self_containing[0] = self_containing
+        try:
+            Error = RecursionError
+        except NameError:
+            Error = RuntimeError
```
Comment (juliantaylor) [inline numpy/core/tests/test_multiarray.py]: should also have a `python 2` comment

**Comment by juliantaylor** [comment]
thanks

**Comment by eric-wieser** [comment]
For some reason, this doesn't fix `a in a` - I can't work out where the recursion is happening, as `Py_RichCompare` is supposed to already call` Py_EnterRecursiveCall`

**Comment by juliantaylor** [comment]
`DEPRECATE` in `arrayobject.c:1411` triggers another enter to the recursion checking. Once one of these errors has been thrown one must abort immediately.

**Comment by eric-wieser** [comment]
Can we remove that deprecation in 1.14?

**Comment by juliantaylor** [comment]
probably not, as far as I recall that was one that caused enough noise to come up on the mailing list, but then I am of the opinion that deprecations are to almost never be removed, others might disagree.

**Comment by eric-wieser** [comment]
> then I am of the opinion that deprecations are to almost never be removed

Well in that case, the deprecation message shouldn't have been "this will raise an error in the future" ;). But fair enough, perhaps 1.14 is too soon

**Pull request merged**